clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // #USN_L2
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2); // Used in the kernel name after builtin
+
+  // kernel7 - expect error
+  // Same as kernel6, except make l6 part of naming kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda to the kernel throws an error because the first
+  // lambda is included in the signature of the second lambda, hence attempting
+  // to change the mangling of the kernel
+  auto l7 = []() { return 1; };
+  auto l8 = [](decltype(l7) *derp = nullptr) { return 2; };
+  constexpr const char *l7_output = __builtin_unique_stable_name(l7); // #USN_l7
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@#kernel7call{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel7>(l8); // #kernel7call
+
+  // kernel8 and kernel9 - expect error
+  // Make two lambdas
+  // Call the builtin on the second of the two lambdas (in a constexpr context)
+  // Within a constexpr if, pass the first lambda to a kernel in the true branch,
+  // pass the second lambda to a kernel in the false branch
+  // Test that passing a lambda to the unique stable name builtin and passing it
+  // to a kernel called with an if constexpr branch causes a diagnostic on the
+  // kernel9 use due to the change in the results to the stable name
+  auto l9 = []() { return 1; };
+  auto l10 = []() { return 2; };
+  constexpr const char *l10_output = __builtin_unique_stable_name(l10); // #USN_l10
+  if constexpr (1) {
+    kernel_single_task<class kernel8>(l9);
+  } else {
+    // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+    // expected-note@#USN_l10{{'__builtin_unique_stable_name' evaluated here}}
+    // expected-note@+1{{in instantiation of function template specialization}}
+    kernel_single_task<class kernel9>(l10);
+  }
+
+  // kernel6 - expect no error
+  // Make a lambda
+  // Make another lambda that captures the first one
+  // Call the builtin on the first lambda (in a constexpr context)
+  // Pass the second lambda to a kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda capturing the first one to the kernel does not
+  // throw an error because the first lambda is not involved in naming the
+  // kernel i.e., the mangling does not change
```

Comment:
```suggestion
  // kernel6 - expect no error
  // Tests that passing a lambda to the unique stable name builtin and then
  // passing a second lambda capturing the first one to the kernel does not
  // throw an error because the first lambda is not involved in naming the
  // kernel i.e., the mangling does not change.
```